File: lib/src/synthesis.cc
#include "synthesis.h"
#include<iostream>
#include<string>
#include<map>
#include<vector>
#include<algorithm>
#include<sstream>
#include<fstream>

std::map<char, char> Organism::rna_matrix = {
    {'A', 'U'},
    {'T', 'A'},
    {'G', 'C'},
    {'C', 'G'}
};

std::string Organism::get_name() const {
    return name_;
}

std::string Organism::get_rna() const {
    return rna_;
}

std::string Organism::get_output() const {
    std::stringstream oput;
    oput << "RNA sequence " << organism_number_ + 1 << ": " << rna_ << "\n\n";
    if (!is_sequenced_) {
        if (!found_start_) {
            oput << "No start sequence found, cannot sequence\n";
            oput << std::string(20, '-') + "\n";
            return oput.str();
        }
        else {
            oput << "No end sequence found, cannot sequence\n";
            oput << std::string(20, '-') + "\n";
            return oput.str();
        }

    }
    oput << "Found start sequence at " << start_position_ << std::endl;
    oput << "Found end sequence " << end_codon_ << "\n\n";
    for (size_t x = 0; x < separated_rna_.size(); x++) 
        oput << separated_rna_[x] << " ";
    oput << "\n\n" << std::string(20, '-') + "\n\n";
    return oput.str();
 }

std::vector<std::string> Organism::get_separated_rna() const {
    return separated_rna_;
} 

void Organism::given_rna() {
    rna_ = dna_;
} 

void Organism::translate() {
    for (size_t i=0;  i<dna_.length(); i++) 
        rna_ += rna_matrix[dna_[i]];
}

void Organism::sequence_protein() {
    char start_sequence[] = "AUG"; 
    
    start_position_ = rna_.find(start_sequence);
    if (start_position_ == std::string::npos) {
        return;
    } else { found_start_ = 1; }
    std::string rna_tmp = rna_.substr(start_position_);
    int len = rna_tmp.length();
    int codon_count = len/3;
    rna_tmp = rna_tmp.substr(0, codon_count*3);
    separated_rna_.push_back("AUG");
     for (int i = 1; i  < codon_count; i++) {
        std::string codon = rna_tmp.substr(3*i, 3);
        if  (codon == "UAA" || codon  == "UAG" || codon == "UGA") {
            separated_rna_.push_back(codon);
            end_codon_ = codon;
            is_sequenced_ = 1;
            break;
        }
        else { 
            separated_rna_.push_back(codon);
         }
    }
} 

bool Organism::is_sequenced() { return is_sequenced_; }

std::string compare_rna(const std::vector<Organism> &orgs) {
    auto n = orgs.size();
    
    std::vector<bool> v(n);
    fill(v.begin(), v.end() - n + 2, true);

    std::vector<std::vector<Organism>> permutations;

    do { 
        std::vector<Organism> perm;
        for (size_t i = 0; i < n; ++i) {
            if (v[i]) {
                perm.push_back(orgs[i]);
             }
        } 
        permutations.push_back(perm);
    } while (prev_permutation(v.begin(), v.end()));

    std::stringstream output;

    for (size_t x = 0; x < permutations.size(); x++) {
        int pt_muts = 0;
        output << "Comparison between " << 
          permutations[x][0].get_name() << " and " <<
          permutations[x][1].get_name() << ":" << std::endl;
        std::vector<std::string> rna1 = permutations[x][0].get_separated_rna();
        std::vector<std::string> rna2 = permutations[x][1].get_separated_rna();
        for (size_t i = 0; i < rna1.size(); i++) {
            if (rna1[i] != rna2[i]) {
                output << "The strands differ at codon " <<
                  i + 1 << ": " << rna1[i] << 
                  " vs " << rna2[i] << std::endl;
                for (int j =0; j < 3; j++) {
                    if (rna1[i][j] != rna2[i][j]) {
                        pt_muts++;
                    }
                }
            }
        }
        output << "\tNumber of point mutations: " << pt_muts << "\n\n";
    }
    return output.str();
  } 

std::string calculate_orgs(std::vector<Organism> &orgs, bool rna_given=false) {
    std::vector<Organism> new_orgs;
    std::string output;
    for (size_t x = 0; x<orgs.size(); x++) {
        if (rna_given) 
            orgs[x].given_rna();
        else 
            orgs[x].translate();
        orgs[x].sequence_protein();
        output += orgs[x].get_output();
        if (orgs[x].is_sequenced()) {
            new_orgs.push_back(orgs[x]);
         }
      }
    if (new_orgs.size() > 1) 
        output += compare_rna(new_orgs);
    return output;
 }

void output_info(const std::string &output_string, bool quiet, bool output_file, std::string filename) {
    if (output_file) {
        std::ofstream outfile;
        outfile.open(filename);
        outfile << output_string;
        outfile.close();
    }
    if (quiet) {
        return;
    }
    std::cout << output_string;

}

void default_process(bool test_mode, bool quiet, bool output_file, std::string filename) {
    int x = 0;
    auto times = test_mode ? 10000 : 1; 
    
    do { 
        std::vector<Organism> orgs;
        Organism afri("African", 0, "CTACGTTCATCTGGTCAGAACTGGTTA");
        Organism ande("Andes", 1, "TCACCTACGTCGATCTGGTCAGGACTT");
        Organism anta("Antarctic", 2, "CACCTACGTTGATCCGGTCAGGACTGGTTA");
        
        orgs.push_back(afri);
        orgs.push_back(ande);
        orgs.push_back(anta);

        std::string outinfo = calculate_orgs(orgs);
        output_info(outinfo, quiet, output_file, filename);
        x++;
     } while (x < times); 
} 

void showhelpinfo(char *s) {
  std::cout<<"Usage:   "<< s <<" [-option] [argument]\n"
           <<"option:  "<<"-h  show help information\n"
           <<"         "<<"-d  use default values, useful in testing\n"
           <<"         "<<"-q  quiet\n"
           <<"         "<<"-o  output file\n"
           <<"         "<<"-r  input is rna\n"
           <<"         "<<"-t  enable test mode\n"
           <<"         "<<"-v  show version infomation\n"
           <<"example: "<< s <<" -o output.txt <DNA1> <DNA2> ...\n"<<std::flush;
}

File: src/python/synthesis.py
from sys import argv, exit
from itertools import combinations

VERSION = 0.6

class Organism():
    rna_matrix = { "A": "U",
                   "T": "A",
                   "G": "C",
                   "C": "G" }

    def __init__(self, name, number, dna):
        self.dna_ = dna
        self.organism_number_ = number 
        self.name_ = name
        self.found_start_ = False
        self.is_sequenced_ = False
        self.separated_rna_ = []
    
    def get_name(self):
        return self.name_
    def get_rna(self):
        return self.rna_
    
    def get_output(self):
        oput = ""
        oput += "RNA sequence {}: {}\n\n".format((self.organism_number_ + 1), self.rna_)
        if not self.is_sequenced_:
            if not self.found_start_:
                oput += "No start sequence found, cannot sequence\n"
                oput += "-"*20 + "\n\n" 
                return oput
            else:
                oput += "No end sequence found, cannot sequence\n"
                oput += "-"*20 + "\n\n"
                return oput
        oput += "Found start sequence at {}\n".format(self.start_position_)
        oput += "Found end sequence {}\n\n".format(self.end_codon_)
        oput += " ".join([x for x in self.separated_rna_])
        oput += "\n\n" + "-"*20 + "\n\n"
        return oput

    def get_separated_rna(self):
        return self.separated_rna_
    def is_sequenced(self):
        return self.is_sequenced_
    def given_rna(self):
        self.rna_ = self.dna_
    def translate(self):
        self.rna_ = ''.join([self.rna_matrix[x] for x in self.dna_])
    
    def sequence_protein(self):
        start_seq = "AUG" 
        end_seqs  = ["UAA", "UAG", "UGA"]
        self.start_position_ = self.rna_.find(start_seq)
        if self.start_position_ == -1:
            self.found_start_ = False
            return
        else:
            self.found_start_ = True
        rna_tmp = self.rna_[self.start_position_:]
        length = len(rna_tmp)
        codon_count = length//3
        rna_tmp = rna_tmp[0:codon_count*3]
        self.separated_rna_.append("AUG")
        for i in range(1,codon_count):
            codon = rna_tmp[ 3*i : 3*i+3 ]
            if codon in end_seqs:
                self.separated_rna_.append(codon)
                self.end_codon_ = codon
                self.is_sequenced_ = True
                break
            else:
                self.separated_rna_.append(codon)

def compare_rna(orgs):
    oput = ""
    for x in combinations(orgs, 2):
        pt_muts = 0
        oput += "Comparison between {} and {}:\n".format(x[0].get_name(), x[1].get_name())
        for pos, codon in enumerate(x[0].get_separated_rna()):
            if codon != x[1].get_separated_rna()[pos]:
                oput += "The strands differ at codon {}: {} vs {}\n".format(
                        (pos+1), 
                        ''.join(x[0].get_separated_rna()[pos]), 
                        ''.join(x[1].get_separated_rna()[pos]))
                for acid in range(0,3):
                    if x[0].get_separated_rna()[pos][acid] != x[1].get_separated_rna()[pos][acid]:
                        pt_muts = pt_muts + 1
        oput += "\tNumber of point mutations: {}\n\n".format(pt_muts)
    return oput

def calculate_orgs(orgs, rna_given=False):
    new_orgs = []
    output = ""
    for x in range(0,len(orgs)):
        if rna_given:
            orgs[x].given_rna()
        else:
            orgs[x].translate()
        orgs[x].sequence_protein()
        output += orgs[x].get_output()
        if (orgs[x].is_sequenced()):
            new_orgs.append(orgs[x])
    if len(new_orgs) > 1:
        output += compare_rna(new_orgs)
    return output 

def output_info(output_string, quiet, output_file, filename):
    if output_file:
        outfile = open(filename, 'w')
        outfile.write(output_string)
        outfile.close()
    if quiet:
        return
    print(output_string)

def default_process(test_mode, quiet, output_file, filename):
    x = 0;
    times = 10000 if test_mode else 1 
    
    while True:
        orgs = []
        afri = Organism("African", 0, "CTACGTTCATCTGGTCAGAACTGGTTA")
        ande = Organism("Andes", 1, "TCACCTACGTCGATCTGGTCAGGACTT")
        anta = Organism("Antarctic", 2, "CACCTACGTTGATCCGGTCAGGACTGGTTA")
        
        orgs.append(afri)
        orgs.append(ande)
        orgs.append(anta)

        outinfo = calculate_orgs(orgs)
        output_info(outinfo, quiet, output_file, filename)
        
        x = x + 1
        if x == times:
            break

def showhelpinfo(script_name):
    print("Usage:   " + script_name + " [-option] [argument]\n" 
         +"option:  " + "-h  show help information\n" 
         +"         " + "-d  use default values, useful in testing\n" 
         +"         " + "-q  quiet\n" 
         +"         " + "-o  output file\n" 
         +"         " + "-r  input is rna\n" 
         +"         " + "-t  enable test mode\n" 
         +"         " + "-v  show version infomation\n" 
         +"example: " + script_name + " -o output.txt <DNA1> <DNA2> ...\n")
